### ISSUE TITLE
Too many .sst files with 4.0k

### ISSUE THREAD

**Issue by deepakjois** [dgraph-io/badger#156]
We ran into this while benchmarking. The data was created with the badger-bench/populate script using the following params:

```
./populate --kv badger --valsz 16384 --keys_mil 5 --dir /mnt/data/16kb
```

doing a `du -sh` on the data dir:

```
$ du -sh *.sst
50M     000001.sst
50M     000002.sst
45M     000003.sst
4.0K    000004.sst
4.0K    000005.sst
4.0K    000006.sst
4.0K    000007.sst
4.0K    000008.sst
4.0K    000009.sst
4.0K    000010.sst
4.0K    000011.sst
4.0K    000012.sst
4.0K    000013.sst
```

@manishrjain  was saying that’s too many 4.0K entries.

**Comment by manishrjain** [comment]
I'm curious to know why this happened.

**Comment by srh** [comment]
Well, the behavior doesn't reproduce, and I'd wonder what badger_info has to say about the structure.  .  What are the exact sizes of the 4.0K .sst's?

**Comment by srh** [comment]
I'm getting 166-byte files (which show up as 4.0K under du).

A file gets written if you actually run a benchmark.  So it took more to reproduce than just running the populate script that was mentioned.

An sst file will get dumped with BenchmarkReadRandomBadger, but not with BenchmarkIterateBadgerWithValues or BenchmarkIterateBadgerOnlyKeys.

**Comment by manishrjain** [comment]
A new SST file should be created only when we do a compaction. Does the presence of these files mean we attempted so many compactions, but they somehow failed to complete?

Can you (@srh or @deepakjois) output badger_info on this directory?

**Comment by srh** [comment]
All the sst's are at level 0.  Doing a get operation is causing some memtable to be deemed dirty, or something like that.

**Comment by manishrjain** [comment]
Sounds like a bug to me.

**Comment by srh** [comment]
The reason BenchmarkIterateBadgerWithValues and BenchmarkIterateBadgerOnlyKeys didn't write an .sst file is that they didn't have a `defer bdb.Close()` line.  A short-run solution to the benchmarking problem is to omit those lines (if turning on compaction doesn't suit you).

**Comment by manishrjain** [comment]
Interesting.

But, if there are no new keys added, then what are we writing to Level 0?

**Comment by srh** [comment]
At the least we're writing !badger!head.  I don't yet know if we're writing anything else.

**Comment by manishrjain** [comment]
Ah.. we should only write a new file if we have valid keys in the memtable. This should be an easy fix then.

**Comment by srh** [comment]
Actually I'm mistaken.  We aren't writing any keys.  In other circumstances we might write just `!badger!head`.  If it's just read-only, then `s.vptr` never gets updated, which means it's zero, which means we don't write `!badger!head`.

So 166 is simply the size of an empty table file.  In any case this is a simple fix.

**Comment by srh** [comment]
Fixed by #160.  In master as of afd5a33.
